### DISCUSSION TITLE
Pagination with session?

### DISCUSSION THREAD

**Discussion by wimski** [livewire/livewire#9081] [1. Help] [answered]
### Livewire version

4.1.4

### Laravel version

12.50.0

### Which PHP version are you using?

PHP 8.4

### What is the problem?

Can pagination be combined with the session attribute?

I have multiple Livewire components that function as tabs in an application using `wire:navigate`. To achieve the sense of persistence, I heavily use the `#[Session]` attribute (usually in combination with `#[Url]`) on properties that deal with user input (search and filter mostly for table data). This all works fine, but I would also like to keep the current pagination page in session, so the user really returns to the exact same view when switching back and forth between tabs.

The query string portion is created dynamically for pagination within the `HandlesPagination` trait. Is there any way to also add session behavior to this?

### Code snippets

.

### Wirebox URL

https://wirebox.app/b/x22w0

### How do you expect it to work?

.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my help request will be removed if I haven't met the criteria above.

**Comment by lrljoe** [comment]
You need to share an example, not just a link to Google

**Reply by wimski** [reply]
It was more a general question if the session functionality can be applied to a paginator somehow.

I'm trying to create a wirebox example now, but the query string parameters do not get added to the URL. Even if I manually add a query string parameter to the URL, it is immediately removed (although the value is applied to the component). This makes it hard to create an example for this specific case.

**Reply by lrljoe** [reply]
Even if you create a Wire Box example that doesn't fully work in Wire Box, it's better than basically telling people who are trying to help to "Google It"

**Reply by wimski** [reply]
> basically telling people who are trying to help to "Google It"

That wasn't my intention. I just filled in 'a' URL in a mandatory field.

I've updated the discussion with new versions and a wirebox URL, which is only the setup I would like to apply session behavior to. The question still remains: how?

~Unrelated (might make a new topic): the pagination info in my wirebox demo is incorrect saying _"Showing 1 to 100 of 100 results"_ instead of _"Showing 1 to 5 of 100 results"_.~

**Comment by wimski** [comment]
@joshhanley since you were so helpful with https://github.com/livewire/livewire/discussions/10000 and this is (somewhat) related, you might want to take a look at this as well :blush:

**Comment by joshhanley** [comment, marked as answer]
@wimski you can achieve this by seeding the request query bag in `mount()` and writing to session via the `updatedPaginators` hook:

```php
use Livewire\Component;
use Livewire\WithPagination;

class MyComponent extends Component
{
    use WithPagination;

    public function mount()
    {
        if (! request()->query->has("page") && session()->has("my-page")) {
            request()->query->set("page", session("my-page"));
        }
    }

    public function updatedPaginators($page, $pageName)
    {
        session()->put("my-page", $page);
    }

    // ...
}
```

The paginator initialises lazily during render — it reads from `request()->query("page", 1)`. By seeding the query bag in `mount()`, the session value is picked up as if it came from the URL, while the normal initialisation flow (including URL tracking) still runs. Priority ends up being: URL param > session > default (1).

**Important:** don't set `$this->paginators["page"]` directly in `mount()` — that skips the URL hook registration and pagination links will stop updating the URL.

**Reply by wimski** [reply]
Thanks! This seems to do the trick. I've wrapped it in a trait for ease of use:
```php
use Illuminate\Contracts\Session\Session;
use Illuminate\Http\Request;
use Livewire\Features\SupportSession\BaseSession;
use Livewire\WithPagination;

trait WithSessionPagination
{
    use WithPagination;

    protected Session $session;

    /**
     * @return string
     */
    abstract function getName();

    /**
     * @return list<string>
     */
    public function getSessionablePageNames(): array
    {
        return ['page'];
    }

    public function bootWithSessionPagination(Session $session): void
    {
        $this->session = $session;
    }

    public function mountWithSessionPagination(Request $request): void
    {
        foreach ($this->getSessionablePageNames() as $pageName) {
            $this->initPaginatorSession($request, $pageName);
        }
    }

    protected function initPaginatorSession(Request $request, string $pageName): void
    {
        if ($request->query->has($pageName)) {
            $this->setPaginatorSession($pageName, (int) $request->query->get($pageName));
            return;
        }

        $sessionKey = $this->makePaginatorSessionKey($pageName);

        if (! $this->session->has($sessionKey)) {
            return;
        }

        $request->query->set($pageName, $this->session->get($sessionKey));
    }

    public function updatedPaginators(int $page, string $pageName): void
    {
        if (! in_array($pageName, $this->getSessionablePageNames())) {
            return;
        }

        $this->setPaginatorSession($pageName, $page);
    }

    protected function setPaginatorSession(string $pageName, int $page): void
    {
        $this->session->put($this->makePaginatorSessionKey($pageName), $page);
    }

    /**
     * @see BaseSession::key()
     */
    protected function makePaginatorSessionKey(string $pageName): string
    {
        return 'lw' . crc32($this->getName() . $pageName);
    }
}
```
